### ISSUE TITLE
Exception messages are not being shown in the console

### ISSUE THREAD

**Issue by negora** [konsoletyper/teavm#505]
Currently, when an application fails, TeaVM throws a native `Error` with the message `Java exception thrown`, and then, fills the stack trace.

Having the stack trace is really useful while developing, of course. But, in a production environment, where code has been optimized and minified, it's also important to count with the original chain of exceptions. In addition to their names, I consider that the messages are also very important, since they contain information about the context in which the exceptions occurred.

For example, if an element has not been found while parsing a response from a server, I want to know which element ID has caused the failure. For this, maybe TeaVM could concatenate the messages of the original exceptions in the message of the `Error`:
```
Error: Java exception thrown.
  foo.bar.ResponseParsingException: The response could not be parsed.
    foo.bar.ElementNotFoundException: The element with ID "age" was not found.
```

If you consider that the new-line characters are not log-friendly, the messages could be separated with another delimiter.

Thank you!

**Comment by ScraM-Team** [comment]
I agree, more information during exceptions would be helpful.  With larger REST payloads, it is sometimes difficult to know which field caused issues.

**Comment by negora** [comment]
I've come to a temporary workaround for this. I've created a small JS script that registers an event listener for unhandled exceptions. This listener runs through the chain of Java exceptions to collect their messages, and prints them to the browser console, one per line, using the error log level.

To convert each Java `String` to a native JS `string`, I've copied and refactored the `$rt_ustr()` function from the `classes.js` file.

This workaround has the following flaws:

1. I've not been able to recover the exception names. Either they're not retained by TeaVM or I've not been able to look for in the right place.

1. This script breaks if you minify your code, because the names of the properties are completely changed.

I've attached the script to this message, in case that other people find it useful: [unhandled-exception-catcher.zip](https://github.com/konsoletyper/teavm/files/4770343/unhandled-exception-catcher.zip).

**Comment by stikdragon** [comment]
My team has recently decided to switch our development to TeaVM, and for the most part we're really liking it, but this exception handling behaviour is probably one of our biggest bugbears.   What's the rationale behind hiding the exception message in this way?   As far as I can tell it would be a very simple thing for it to put the actual message in there instead of just "Java exception thrown"

@negora I think TeaVM has changed enough since you make this post that your script no longer works, however it gave me enough insight to find something that works well enough for me:

```js
(function() {
	window.addEventListener("error", function(event) {
		console.error(main.javaException(event.error).$message.toString());
		return false;
	});
})();
```
It just dumps out the final exception's message, which in 99% of cases is enough to let me know what Dumb Thing it is that I have just done

**Comment by konsoletyper** [comment]
> What's the rationale behind hiding the exception message in this way

@stikdragon simply the Java spec. `Throwable.printStackTrace()` is expected to show exceptions in terms of Java stack traces. You can also extract StackTraceElement array from exception, or from current thread, call `toString` on them and get exactly the same result. JavaScript stack traces are completely different. Moreover, there's no JavaScript spec for stack trace format, so different engines just produce different strings. Transforming JavaScript stack traces to Java stack traces also requires large tables, which increase the size of JavaScript and memory consumption; parsing these tables takes some time, which increases boot time. Printing exception to console is a security risk, since it can uncover implementation details to a hacker. Having a deobfuscation table loaded by client is double security risk. So the decision was not to emulate Java stack traces in production environment, since usually stack traces are only important for developers.

If you are [developing using IDEA](https://teavm.org/docs/tooling/idea.html), you can use development server, which handles stack traces properly. Sorry, it's currently for IDEA, but you can borrow the same code to run from other environments. I'm considering of a way to run development server from Gradle.

There's another use case: collect stack traces from users and deobfuscate them on the server, like we do in [CoSpaces Edu](https://cospaces.io/edu/). However, this requires some extra setup and extra documentation. I could have documented this, but I'm also thinking of a way to turn TeaVM into my own business that brings money, so perhaps things like these could be a good candidates for consulting or non-free documentation.

**Comment by konsoletyper** [comment]
@stikdragon you can also take a look at `JsExceptions` class from JSO library, which helps to convert between JavaScript and Java exception

**Comment by stikdragon** [comment]
Thanks for the quick response.  I understand that converting between JS and Java stack traces is non-trivial and your reasoning is sound (although I tend to disagree about the security implications, this code runs on the client and obfuscation is only going to slow a bad actor down, not stop them), but my issue is with the exception message itself.

Eg: I add a click handler to a button, and i've written an off-by-one error in the function it calls that causes an `IndexOutOfBoundsException` to get thrown.  Clicking the button causes just the message "Java Exception thrown" to appear in the console, and now I have to spend way more time trying to find out what the actual problem was.   If it had said "Java Exception thrown: IndexOutOfBoundsException: Index 5 out of bounds for length 5" then I would immediately have guessed at the cause of my problem.

Currently my approach is to set an exception breakpoint in Chrome, and then try to work out what's happened from there, but that's hampered by `String` not being implemented as a simple JS string, so it's not a very smooth workflow either.  I suppose you could argue that an uncaught exception is a problem in itself, but wrapping everything in global catch blocks is also a clumsy development experience

In any case, I have an acceptable workaround for this now that I've found the `main.javaException` function (which is maybe related to the `JsExceptions` class you mentioned?).  Just wanted to hear your thoughts on it

Cheers

**Comment by konsoletyper** [comment]
> but wrapping everything in global catch blocks is also a clumsy development experience

I see. But in case you don't catch exception, it will be caught by JS interpreter and printed to console. According to what I see, it's not your case. You have this exception caught somewhere in Java code and logged by `Throwable.printStackTrace`. As I already said, the development experience is smooth when you use developer server.

**Comment by konsoletyper** [comment]
@stikdragon So, it's not "exception stack", but "exception message". Well, the reason is following. TeaVM cares a bit more about in-JVM performance. This means that whatever Java code exists, it should be emulated as fast as possible. TeaVM trades this for a bit worse JS interop performance and sometimes developer experience. GWT has different approach than that, for being as close to JS as possible it trades in-JVM performance (and sometimes semantics). This has the following consequence: TeaVM does not represent basic types (primitive wrappers, arrays, String, Exception, Object, List) as JavaScript standard objects. For example, it re-implements `java.lang.String` as a fair object that contains `char[]` field (which is represented as `UInt16Array`, wrapped into an object in JS). Thus when you deal with something like `java.lang.String`, `java.lang.Integer` or `java.lang.Throwable`, you don't deal with a native JavaScript object, but with a wrapper. What you seen in the console is a JavaScript Error that represents corresponding Java Throwable object. Error object has only reference to originating Throwable, but does not expose all other properties, due to performance reasons. 10 years ago, when I just started TeaVM, overriding standard Error properties was a bit challenging. I'll look at recent state and if it's possible to override message property lazily, I'll expose message from Throwable to JS Error.

As for stack trace, it's simple here: you can enable source maps generation and Chrome will display it properly.

> but we're currently pretty ingrained into the world of Eclipse

You can run devserver programmatically, use [DevServer](https://github.com/konsoletyper/teavm/blob/07d56f1b0b0b63e2692cdecbd137fc81f6ec95cf/tools/devserver/src/main/java/org/teavm/devserver/DevServer.java) class. This should be trivial. You can write a simple command-line tool for your particular needs or implement Eclipse plugin that just wraps this class. You can also build [CLI](https://github.com/konsoletyper/teavm/tree/07d56f1b0b0b63e2692cdecbd137fc81f6ec95cf/tools/cli) from sources, there's already CLI tool for running dev server.

**Comment by konsoletyper** [comment]
@stikdragon I implemented proper mapping of messages from Java Throwable to JS Error in [0f8d36a](https://github.com/konsoletyper/teavm/commit/0f8d36a0807c9dc4d4475fdaab71542dbcc6028c). Commit also maps causes, however currently only Firefox devtools support this feature.
